python/lsst/ap/pipe/ap_pipe.py

```diff
@@ -802,8 +804,10 @@ def _doDiffIm(processed_repo, dataId, templateType, template, diffim_repo):
     log.info('Running ImageDifference...')
     if not os.path.isdir(diffim_repo):
         os.mkdir(diffim_repo)
-    diffim_result = ImageDifferenceTask.parseAndRun(args=args, config=config, doReturnResults=True)
-    diffim_metadata = diffim_result.resultList[0].metadata
+    ImageDifferenceTask.parseAndRun(args=args, config=config, doReturnResults=True)
+    butler = dafPersist.Butler(inputs=diffim_repo)
+    metadataType = ImageDifferenceTask()._getMetadataName()
+    diffim_metadata = butler.get(metadataType, dataId_dict)
```

Comment:
I found this a bit confusing, primarily because I'm not sure I understand what's supposed to be in `dataId`.

The example in the docstring for this function looks as though it's the sort of string you might pass on the command line, potentially specifying multiple data units (e.g. `ccd=1^2^3^4 visit=5^6^7^8`). But the way it's treated in the code implies that it's actually a fully-specified data ID.

This makes a difference, since you're handing it off to `parseAndRun`, which will iterate over all the data units if there's more than one. And if that happens, I think your logic above will break (since you're not requesting a well-qualified unit of metadata from the Butler). (Mind you, other logic in the function will probably have broken first...).

All of this makes me wonder why we're calling `parseAndRun()` here, rather than just `run()` (which is what we do in `_doProcessCcd()` above): the latter seems much more straightforward and easy to reason about.

Of course, introducing `parseAndRun()` is not a new innovation in this changeset.

Reply:
You are correct that the code will break if there are multiple data units specified. However, other functions (such as `_doProcessCcd`) will break under those circumstances as well. IMO this is the main benefit of making `ap_pipe` a command-line task: the dataID-parsing code in `pipe_base` is so tightly coupled to the rest of `ArgumentParser` that using the whole thing is about the only option.

What I really want to say is "give me all the metadata generated during this run", but as far as I can tell the only way to do this is to iterate over the expanded set of data units.

I would argue that `parseAndRun` is what we should be calling, because that would let us support parallel processing of multiple datasets once the too-specific dataID handling is fixed. Meredith did not use it for the other top-level tasks because she said that something was incompatible with `parseAndRun`... I think it might have been an `obs_decam` issue, but I unfortunately don't remember.